File: src/main/java/me/arasple/mc/trmenu/actions/BaseAction.java
package me.arasple.mc.trmenu.actions;

import me.arasple.mc.trmenu.actions.option.ActionOption;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryEvent;

import java.util.HashMap;

/**
 * @author Arasple
 * @date 2019/10/2 9:23
 */
public abstract class BaseAction {

    private String command;
    private HashMap<ActionOption, String> options;

    public BaseAction(String command, HashMap<ActionOption, String> options) {
        this.command = command;
        this.options = options;
    }

    /**
     * 执行动作时的方法
     *
     * @param player 玩家
     * @param e      容器事件
     */
    public abstract void onExecute(Player player, InventoryEvent e);

    public String getCommand() {
        return command;
    }

    public void setCommand(String command) {
        this.command = command;
    }

    public HashMap<ActionOption, String> getOptions() {
        return options;
    }

    public void setOptions(HashMap<ActionOption, String> options) {
        this.options = options;
    }

}

File: src/main/kotlin/me/arasple/mc/trmenu/module/internal/script/kether/BaseAction.kt
package me.arasple.mc.trmenu.module.internal.script.kether

import me.arasple.mc.trmenu.module.display.MenuSession
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import openapi.kether.QuestAction
import openapi.kether.QuestContext
import taboolib.module.kether.ScriptContext
import java.util.concurrent.CompletableFuture

/**
 * @author Arasple
 * @date 2021/1/29 10:18
 */
abstract class BaseAction<T> : QuestAction<T>() {

    internal val completableFuture: CompletableFuture<Void> = CompletableFuture.completedFuture(null)

    fun CommandSender.session(): MenuSession? {
        return if (this is Player) MenuSession.getSession(this) else null
    }

    fun QuestContext.Frame.viewer(): Player {
        return ((this.context() as ScriptContext).sender ?: throw RuntimeException("No viewer found")).cast()
    }

}